addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/DSMRHandlerFactory.java

```diff
@@ -121,18 +121,15 @@ private synchronized void registerDiscoveryService(DSMRBridgeHandler bridgeHandl
     @Override
     protected synchronized void removeHandler(ThingHandler thingHandler) {
         if (thingHandler instanceof DSMRBridgeHandler) {
-            ServiceRegistration<?> serviceReg = this.discoveryServiceRegs.get(thingHandler.getThing().getUID());
-
+            ServiceRegistration<?> serviceReg = this.discoveryServiceRegs.remove(thingHandler.getThing().getUID());
             if (serviceReg != null) {
                 DSMRMeterDiscoveryService service = (DSMRMeterDiscoveryService) getBundleContext()
                         .getService(serviceReg.getReference());
+                serviceReg.unregister();
                 if (service != null) {
                     service.unsetLocaleProvider();
```

Comment:
@cweitkamp In this situation, it is not OSGi DS that does the lifecycle handling for the services, but we have to do it - so just like DS does the reference injection and removal, we should do the very same (so that the service itself can rely on being called that way).

Reply:
@kaikreuzer I got it. Thanks for explanation.